src/core/vmm.d

```diff
@@ -0,0 +1,489 @@
+    {
+        return VirtualAlloc(base, size, MEM_RESERVE,
+            PAGE_READWRITE);
+    }
+
+    bool osCommit(void* ptr, size_t size)
+    {
+        ptr = VirtualAlloc(ptr, size, MEM_COMMIT, PAGE_READWRITE);
+        return ptr != null;
+    }
+
+    bool osReset(void* ptr, size_t size)
+    {
+        return VirtualFree(ptr, size, MEM_RESET) != 0;
+    }
+
+    bool osDecommit(void* ptr, size_t size)
+    {
+        return VirtualFree(ptr, size, MEM_DECOMMIT) != 0;
+    }
+
+    bool osProtect(void* ptr, size_t size, MemoryAccess access)
+    {
+        uint oldProt;
+        return VirtualProtect(ptr, size, protectionFlags(access), &oldProt) != 0;
+    }
+
+    bool osLockPages(void* ptr, size_t size)
+    {
+        return VirtualLock(ptr, size) != 0;
+    }
+
+    bool osUnlockPages(void* ptr, size_t size)
+    {
+        return VirtualUnlock(ptr, size) != 0;
+    }
+}
+else version(Posix)
+{
+    version(linux)
+        import core.sys.linux.sys.mman;
+    else version(FreeBSD)
+        import core.sys.freebsd.sys.mman;
+    else version(OSX)
+        import core.sys.darwin.sys.mman;
+    else
+        import core.sys.posix.sys.mman;
+
+    uint protectionFlags(MemoryAccess access)
+    {
+        if(access == MemoryAccess.none)
+            return PROT_NONE;
+        uint protFlags;
+        if(access & MemoryAccess.read)
+            protFlags |= PROT_READ;
+        if(access & MemoryAccess.write)
+            protFlags |= PROT_WRITE;
+        if(access & MemoryAccess.execute)
+            protFlags |= PROT_EXEC;
+        return protFlags;
+    }
+
+    uint memoryFlags(uint options)
+    {
+        //TODO: no huge pages yet, the whole procedure isn't trivial
+        version(linux)
+            return MAP_PRIVATE | MAP_ANONYMOUS;
+        else
+            return MAP_PRIVATE | MAP_ANON;
+    }
+
+    void* osMap(void* base, size_t nbytes, MemoryAccess access, uint options)
+    {
+        uint memFlag = memoryFlags(options) | (base == null ? 0 : MAP_FIXED);
+        void* p = mmap(base, nbytes, protectionFlags(access), memFlag, -1, 0);
+        return p == MAP_FAILED ? null : p;
+    }
+
+    bool osUnmap(void *base, size_t nbytes)
+    {
+        return munmap(base, nbytes) == 0;
+    }
+
+    void* osReserve(void* base, size_t size)
+    {
+        return osMap(base, size, MemoryAccess.readWrite, MemoryOptions.none);
+    }
+
+    bool osCommit(void* ptr, size_t size)
+    {
+        ptr = roundDownToPage(ptr);
+        size = roundUpToPage(size);
+        version(linux)
+            return madvise(ptr, size, MADV_WILLNEED) == 0;
+        else
+            return posix_madvise(ptr, size, POSIX_MADV_WILLNEED) == 0;
+    }
+
+    bool osDecommit(void* ptr, size_t size)
+    {
+        ptr = roundDownToPage(ptr);
+        size = roundUpToPage(size);
+        version(linux)
+            return madvise(ptr, size, MADV_DONTNEED) == 0;
+        else
+            return posix_madvise(ptr, size, POSIX_MADV_DONTNEED) == 0;
```

Comment:
Well, I think it's clear by now that the module should return some kind of ErrorCode instead of bool on most operations. The tricky one is allocate though that returns `void[]`.